game/kernel/asm_funcs_arm64.s

```diff
@@ -0,0 +1,262 @@
+;; GOAL Runtime assembly functions. These exist only in the arm64 version of GOAL.
+;; - https://developer.apple.com/documentation/xcode/writing-arm64-code-for-apple-platforms#Pass-arguments-to-functions-correctly
+;; - https://en.wikipedia.org/wiki/Calling_convention#ARM_(A64)
+;; - https://student.cs.uwaterloo.ca/~cs452/docs/rpi4b/aapcs64.pdf
+;; - s16–s31 (d8–d15, q4–q7) must be preserved
+;; - s0–s15 (d0–d7, q0–q3) and d16–d31 (q8–q15) do not need to be preserved
+;; - https://devblogs.microsoft.com/oldnewthing/20220728-00/?p=106912
+;; - ;; - https://courses.cs.washington.edu/courses/cse469/19wi/arm64.pdf
+
+.text
+
+;; Call C++ code on arm64 systems, from GOAL.
+;; Following the macOS documentation which mostly aligns with standard arm64
+.global _arg_call_arm64
+.align 4
+_arg_call_arm64:
+  ldr x8, [sp], #16
+
+  ; Putting an exclamation point after the close-bracket 
+  ; means that the calculated effective address is written back to the base register. (pre-indexing)
+  stp q15, q14, [sp, #-32]!
+  stp q13, q12, [sp, #-32]!
+  stp q11, q10, [sp, #-32]!
+  stp q9, q8, [sp, #-32]!
+
+  blr x8
+
+  ldp x8, x9, [sp], #32
+  ldp x10, x11, [sp], #32
+  ldp x12, x13, [sp], #32
+  ldp x14, x15, [sp], #32
+
+  ret
+
+
+;; Call C++ code on arm64 systems, from GOAL. 
+;; 
+;; Put arguments on the stack and put a pointer to this array in the first arg.
+;; this function pushes all 8 OpenGOAL registers into a stack array.
+;; then it calls the function pointed to by x0 (RAX in x86) with a pointer to this array.
+;; it returns the return value of the called function.
+.global _stack_call_arm64
+.align 4
+_stack_call_arm64:
+  ldr x8, [sp], #16
+
+  stp q15, q14, [sp, #-32]!
+  stp q13, q12, [sp, #-32]!
+  stp q11, q10, [sp, #-32]!
+  stp q9, q8, [sp, #-32]!
+
+  ; create stack array of arguments
+  ; arg 7 (R11 in x86)
+  ; arg 6 (R10 in x86)
+  ; arg 5 (R8 in x86)
+  ; arg 4 (R8 in x86)
+  ; arg 3 (RCX in x86)
+  ; arg 2 (RDX in x86)
+  ; arg 1 (RSI in x86)
+  ; arg 0 (RDI in x86)
+  stp x7, x6, [sp, #-16]!
+  stp x5, x4, [sp, #-16]!
+  stp x3, x2, [sp, #-16]!
+  stp x1, x0, [sp, #-16]!
+
+  ; set first argument
+  mov x19, sp
+  ; call function
+  blr x8
+  ; restore arguments
+  ldp x0, x1, [sp], #16
+  ldp x2, x3, [sp], #16
+  ldp x4, x5, [sp], #16
+  ldp x6, x7, [sp], #16
+
+  ldp x8, x9, [sp], #32
+  ldp x10, x11, [sp], #32
+  ldp x12, x13, [sp], #32
+  ldp x14, x15, [sp], #32
+
+  ; return!
+  ret
+
+;; Call c++ code through mips2c.
+;; GOAL will call a dynamically generated trampoline.
+;; The trampoline will have pushed the exec function and stack offset onto the stack
+.global _mips2c_call_arm64
+.align 4
+_mips2c_call_arm64:
+  ;; TODO - this is really weird using half an XMM, this makes the arm assembly
+  ;; more difficult - this probably isn't required for arm?
```

Comment:
yep, you're right. I only used `xmm0` because all the x86 GPRs were used and it was convenient.